### PR TITLE
feat: Update widget swap, refactor UI components to use widget state 

### PR DESCRIPTION
**Pull request by ianlapham**
- major changes to state structure in widget state 
- refactor majority of files to use real derived swap data 
- in progress - refactor still needed on a couple more components - will break in cosmos

### CODE REVIEW

File: src/lib/components/Swap/Status.fixture.tsx [hunk 1]
```diff
@@ -16,8 +16,8 @@ function Fixture() {
   })
   useEffect(() => {
     setTransaction({
-      input: { token: ETH, value: 1 },
-      output: { token: DAI, value: 4200 },
+      input: Ether,
```
Comment (JFrankfurt) [inline src/lib/components/Swap/Status.fixture.tsx:19]: is Ether defined here?

File: src/hooks/useAllCurrencyCombinations.ts [hunk 2]
```diff
@@ -6,7 +6,9 @@ import { ADDITIONAL_BASES, BASES_TO_CHECK_TRADES_AGAINST, CUSTOM_BASES } from '.
 export function useAllCurrencyCombinations(currencyA?: Currency, currencyB?: Currency): [Token, Token][] {
   const chainId = currencyA?.chainId
 
-  const [tokenA, tokenB] = chainId ? [currencyA?.wrapped, currencyB?.wrapped] : [undefined, undefined]
+  const [tokenA, tokenB] = useMemo(() => {
```
Comment (zzmp) [inline src/hooks/useAllCurrencyCombinations.ts:9]: This should be unnecessary, because you're never using the array returned by `useMemo`, and `tokenA === currencyA?.wrapped`.

You should revert this.

File: src/hooks/useAllV3Routes.ts [hunk 3]
```diff
@@ -66,10 +66,15 @@ export function useAllV3Routes(
   const { chainId } = useActiveWeb3React()
   const { pools, loading: poolsLoading } = useV3SwapPools(currencyIn, currencyOut)
 
+  useEffect(() => {
```
Comment (zzmp) [inline src/hooks/useAllV3Routes.ts:69]: debug: revert this

File: src/hooks/useColor.ts [hunk 4]
```diff
@@ -11,7 +11,7 @@ function URIForEthToken(address: string) {
   return `https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/${address}/logo.png`
 }
 
-async function getColorFromToken(token: Token): Promise<string | null> {
+export async function getColorFromToken(token: Token): Promise<string | null> {
```
Comment (zzmp) [inline src/hooks/useColor.ts:14]: revert

File: src/lib/components/Input.tsx [hunk 5]
```diff
@@ -130,7 +125,7 @@ const integerEnforcer = (nextUserInput: string) => {
   return null
 }
 export const IntegerInput = forwardRef(function IntegerInput(props: NumericInputProps, ref) {
-  return <NumericInput pattern="^[0-9]*$" enforcer={integerEnforcer} ref={ref as any} {...props} />
+  return <NumericInput pattern="^[0-9]*[.,]?[0-9]*$" enforcer={integerEnforcer} ref={ref as any} {...props} />
```
Comment (zzmp) [inline src/lib/components/Input.tsx:128]: `IntegerInput` should only allow integers. Revert this change, and use `DecimalInput` if you need decimals.

File: src/lib/components/Input.tsx [hunk 6]
```diff
@@ -144,8 +139,26 @@ const decimalEnforcer = (nextUserInput: string) => {
   }
   return null
 }
+
 export const DecimalInput = forwardRef(function DecimalInput(props: NumericInputProps, ref) {
-  return <NumericInput pattern="^[0-9]*[.,]?[0-9]*$" enforcer={decimalEnforcer} ref={ref as any} {...props} />
+  return (
+    <NumericInput
+      // universal input options
```
Comment (zzmp) [inline src/lib/components/Input.tsx:146]: These options should be added to the `NumericInput` internal component, not to the `DecimalInput` directly.

Comment (zzmp) [reply src/lib/components/Input.tsx:146]: This is done already on L103. I think you should add maxLength to that list, and otherwise revert this line.

File: src/lib/components/Swap/Input.tsx [hunk 7]
```diff
@@ -27,31 +25,38 @@ interface InputProps {
...
-  const balance = mockCurrencyAmount
+  const { onCurrencySelection, onUserInput } = useSwapActionHandlers()
+  const {
+    currencies: { [Field.INPUT]: inputCurrency },
+    currencyBalances: { [Field.INPUT]: balance },
+    parsedAmounts: { [Field.INPUT]: inputAmount },
+  } = useDerivedSwapInfo()
+
+  const inputAmountUSDC = useUSDCValue(inputAmount)
+  //@TODO - ianlapham - mimic logic from app swap page
```
Comment (zzmp) [inline src/lib/components/Swap/Input.tsx:36]: nit: prefer `TODO(ianlapham): mimic...`

File: src/lib/components/Swap/Output.tsx [hunk 8]
```diff
@@ -33,32 +36,34 @@ interface OutputProps {
...
   const overrideColor = useAtomValue(colorAtom)
-  const dynamicColor = useCurrencyColor(output.token)
-  usePrefetchCurrencyColor(input.token) // extract eagerly in case of reversal
+  const dynamicColor = useCurrencyColor(outputCurrency)
+  usePrefetchCurrencyColor(inputCurrency) // extract eagerly in case of reversal
   const color = overrideColor || dynamicColor
-  const hasColor = output.token ? Boolean(color) || null : false
+  const hasColor = outputCurrency ? Boolean(color) || null : false
 
+  const inputUSDCValue = useUSDCValue(inputAmount)
```
Comment (zzmp) [inline src/lib/components/Swap/Output.tsx:52]: nit: This calls it `inputUSDCValue`, but `Input.tsx` calls it `inputAmountUSDC`. Rename both to `inputUSDC`?

File: src/lib/components/Swap/Output.tsx [hunk 9]
```diff
@@ -68,7 +73,13 @@ export default function Output({ disabled, children }: OutputProps) {
             <Trans>For</Trans>
           </ThemedText.Subhead2>
         </Row>
-        <TokenInput input={output} disabled={disabled} onChangeInput={setValue} onChangeToken={setToken}>
+        <TokenInput
+          currency={outputCurrency}
+          amount={outputAmount}
+          disabled={disabled}
+          onChangeInput={(val) => (val ? onUserInput(Field.OUTPUT, val?.toString()) : null)}
```
Comment (zzmp) [inline src/lib/components/Swap/Output.tsx:80]: I don't think it is possible for `val` to be nullish.

File: src/lib/components/Swap/Settings/MaxSlippageSelect.tsx [hunk 10]
```diff
@@ -78,8 +78,8 @@ export default function MaxSlippageSelect() {
         <InputOption value={custom} onSelect={onInputSelect} selected={maxSlippage === CUSTOM}>
           <DecimalInput
             size={custom === undefined ? undefined : 5}
-            value={custom}
-            onChange={(custom) => setMaxSlippage({ value: CUSTOM, custom })}
+            value={custom?.toString()}
+            onChange={(custom) => (custom ? setMaxSlippage({ value: CUSTOM, custom: parseFloat(custom) }) : null)}
```
Comment (zzmp) [inline src/lib/components/Swap/Settings/MaxSlippageSelect.tsx:82]: Unless you've changed `setMaxSlippage`, I think we want this to always invoke it, with `custom: custom ? parseFloat(custom) : null`.

File: src/lib/components/Swap/Settings/MockToggle.tsx [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import { Trans } from '@lingui/macro'
 import { useAtom } from 'jotai'
-import { mockTogglableAtom } from 'lib/state/swap'
+import { mockTogglableAtom } from 'lib/state/settings'
```
Comment (zzmp) [inline src/lib/components/Swap/Settings/MockToggle.tsx:3]: This file is only here so we don't have to recreate work on toggles later on. Thanks for keeping it up to date!

File: src/lib/components/Swap/Settings/TransactionTtlInput.tsx [hunk 12]
```diff
@@ -25,8 +25,8 @@ export default function TransactionTtlInput() {
         <Input onClick={() => input.current?.focus()}>
           <IntegerInput
             placeholder={TRANSACTION_TTL_DEFAULT.toString()}
-            value={transactionTtl}
-            onChange={(value) => setTransactionTtl(value ?? 0)}
+            value={transactionTtl?.toString()}
+            onChange={(value) => setTransactionTtl(value && typeof value === 'string' ? parseFloat(value) : 0)}
```
Comment (zzmp) [inline src/lib/components/Swap/Settings/TransactionTtlInput.tsx:29]: I think `value ? parseFloat(value) : 0` is simpler.

File: src/lib/components/Swap/Status.fixture.tsx [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import { tokens } from '@uniswap/default-token-list'
-import { SupportedChainId } from 'constants/chains'
-import { nativeOnChain } from 'constants/tokens'
+import { ChainId } from '@uniswap/smart-order-router'
```
Comment (zzmp) [inline src/lib/components/Swap/Status.fixture.tsx:2]: Use `SupportedChainId` from `constants/chains`.

File: src/lib/components/Swap/Status.fixture.tsx [hunk 14]
```diff
@@ -1,6 +1,6 @@
...
+import { WRAPPED_NATIVE_CURRENCY } from 'constants/tokens'
```
Comment (zzmp) [inline src/lib/components/Swap/Status.fixture.tsx:3]: revert

File: src/lib/components/Swap/Status.fixture.tsx [hunk 15]
```diff
@@ -11,7 +11,7 @@ import invariant from 'tiny-invariant'
 import { Modal } from '../Dialog'
 import { StatusDialog } from './Status'
 
-const ETH = nativeOnChain(SupportedChainId.MAINNET)
+const ETH = WRAPPED_NATIVE_CURRENCY[ChainId.MAINNET]
```
Comment (zzmp) [inline src/lib/components/Swap/Status.fixture.tsx:14]: revert. Is there a reason to prefer WRAPPED_NATIVE_CURRENCY over nativeOnChain?

File: src/lib/components/Swap/Summary.fixture.tsx [hunk 16]
```diff
@@ -1,6 +1,5 @@
 import { tokens } from '@uniswap/default-token-list'
-import { SupportedChainId } from 'constants/chains'
-import { nativeOnChain } from 'constants/tokens'
+import { DAI } from 'constants/tokens'
```
Comment (zzmp) [inline src/lib/components/Swap/Summary.fixture.tsx:2]: revert. Use UNI instead of DAI (from L13).

File: src/lib/components/Swap/Summary/Details.tsx [hunk 17]
```diff
@@ -23,31 +26,30 @@ function Detail({ label, value }: DetailProps) {
...
   const details = useMemo((): [string, string][] => {
+    // @TODO ianlapham = update details to pull derived value from useDerivedSwapInfo
     return [
-      [t`Liquidity provider fee`, `${swap.lpFee} ${inputSymbol}`],
-      [t`${integrator} fee`, swap.integratorFee && `${swap.integratorFee} ${inputSymbol}`],
-      [t`Price impact`, `${swap.priceImpact}%`],
-      [t`Maximum sent`, swap.maximumSent && `${swap.maximumSent} ${inputSymbol}`],
-      [t`Minimum received`, swap.minimumReceived && `${swap.minimumReceived} ${outputSymbol}`],
-      [t`Slippage tolerance`, `${swap.slippageTolerance}%`],
+      // [t`Liquidity provider fee`, `${swap.lpFee} ${inputSymbol}`],
```
Comment (zzmp) [inline src/lib/components/Swap/Summary/Details.tsx:41]: Stick a TODO on top of this so it doesn't get forgotten.

File: src/lib/components/Swap/Summary/index.tsx [hunk 18]
```diff
@@ -124,12 +123,12 @@ export function SummaryDialog({ onConfirm }: SummaryDialogProps) {
...
           <DetailsColumn>
             <Column gap={0.5} ref={setDetails} css={scrollbar}>
-              <Details input={input.token} output={output.token} swap={swap} />
+              <Details input={inputCurrency} output={outputCurrency} />
             </Column>
           </DetailsColumn>
           <Estimate color="secondary">
             <Trans>Output is estimated.</Trans>{' '}
-            {swap?.minimumReceived && (
+            {/* {swap?.minimumReceived && (
```
Comment (zzmp) [inline src/lib/components/Swap/Summary/index.tsx:131]: Stick a TODO over this so it's not forgotten.

File: src/lib/components/TokenSelect/useQueriedTokenList.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import useDebounce from 'hooks/useDebounce'
```
Comment (zzmp) [inline src/lib/components/TokenSelect/useQueriedTokenList.ts:1]: merge conflict: delete this file.

File: src/lib/components/TokenSelect/utils.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
+import { Currency, CurrencyAmount, Token } from '@uniswap/sdk-core'
```
Comment (zzmp) [inline src/lib/components/TokenSelect/utils.ts:1]: merge conflict: delete this file

File: src/lib/hooks/trade.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
```
Comment (zzmp) [inline src/lib/hooks/trade.ts:1]: nit: rename to useBestTrade.ts

File: src/lib/hooks/useColor.ts [hunk 22]
```diff
@@ -0,0 +1,85 @@
+import { Currency, Token } from '@uniswap/sdk-core'
```
Comment (zzmp) [inline src/lib/hooks/useColor.ts:1]: merge conflict: delete this file

File: src/lib/components/Swap/Settings/MaxSlippageSelect.tsx (hunk 10 above)
Comment (ianlapham) [reply src/lib/components/Swap/Settings/MaxSlippageSelect.tsx:82]: should we assign back to undefined instead? -> custom ? parseFloat(custom) : undefined

File: src/lib/components/Swap/Status.fixture.tsx (hunk 15 above)
Comment (ianlapham) [reply src/lib/components/Swap/Status.fixture.tsx:14]: Theres no real reason here - i only did this because other places in the app use it. But nativeOnChain is probably more accurate here

File: src/lib/components/Swap/Settings/MaxSlippageSelect.tsx (hunk 10 above)
Comment (zzmp) [reply src/lib/components/Swap/Settings/MaxSlippageSelect.tsx]: Let's do `custom ? parseFloat(custom) : undefined`. It's logically identical to `custom && parseFloat(custom)`, but i think it's more clear to a reader.

File: src/lib/utils/tryParseAmount.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+  try {
+    const typedValueParsed = parseUnits(value, currency.decimals).toString()
+    if (typedValueParsed !== '0') {
+      return CurrencyAmount.fromRawAmount(currency, JSBI.BigInt(typedValueParsed))
+    }
+  } catch (error) {
+    // should fail if the user specifies too many decimal places of precision (or maybe exceed max uint?)
+    console.debug(`Failed to parse input amount: "${value}"`, error)
+  }
+  // necessary for all paths to return a value
```
Comment (zzmp) [inline src/lib/utils/tryParseAmount.ts:19]: Remove comment - unnecessary.

File: src/lib/utils/tryParseAmount.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+  if (!value || !currency) {
+    return undefined
+  }
+  try {
+    const typedValueParsed = parseUnits(value, currency.decimals).toString()
+    if (typedValueParsed !== '0') {
+      return CurrencyAmount.fromRawAmount(currency, JSBI.BigInt(typedValueParsed))
+    }
+  } catch (error) {
+    // should fail if the user specifies too many decimal places of precision (or maybe exceed max uint?)
```
Comment (zzmp) [inline src/lib/utils/tryParseAmount.ts:16]: Try to "verb-ize" comments: `// fails if the user...`

**Comment by ianlapham** [comment]
Closing, stale
